From 082097de69546dc92f52b69e8c4fdc125d9a5d50 Mon Sep 17 00:00:00 2001
From: Keunhong Park <keunhong@codeaurora.org>
Date: Tue, 17 Jul 2012 09:48:31 -0600
Subject: [PATCH] Add dependency errors to UI

The change info block will now display dependency errors if there are any.

Change-Id: Iebea1e06c17aef89d7fad309d92ab10d4683be82
---
 .../gerrit/client/changes/ChangeConstants.java     |    1 +
 .../client/changes/ChangeConstants.properties      |    1 +
 .../client/changes/ChangeDescriptionBlock.java     |    6 ++-
 .../gerrit/client/changes/ChangeInfoBlock.java     |   34 ++++++++++++++++++--
 .../google/gerrit/client/changes/ChangeScreen.java |    3 +-
 .../client/changes/PublishCommentScreen.java       |    2 +-
 .../main/java/com/google/gerrit/client/gerrit.css  |    5 +++
 7 files changed, 45 insertions(+), 7 deletions(-)

diff --git a/gerrit-gwtui/src/main/java/com/google/gerrit/client/changes/ChangeConstants.java b/gerrit-gwtui/src/main/java/com/google/gerrit/client/changes/ChangeConstants.java
index d42992f..c6119ce 100644
--- a/gerrit-gwtui/src/main/java/com/google/gerrit/client/changes/ChangeConstants.java
+++ b/gerrit-gwtui/src/main/java/com/google/gerrit/client/changes/ChangeConstants.java
@@ -99,6 +99,7 @@ public interface ChangeConstants extends Constants {
   String changeInfoBlockCanMerge();
   String changeInfoBlockCanMergeYes();
   String changeInfoBlockCanMergeNo();
+  String changeInfoErrors();
 
   String includedInTableBranch();
   String includedInTableTag();
diff --git a/gerrit-gwtui/src/main/java/com/google/gerrit/client/changes/ChangeConstants.properties b/gerrit-gwtui/src/main/java/com/google/gerrit/client/changes/ChangeConstants.properties
index 8ceb74c..4ed87af 100644
--- a/gerrit-gwtui/src/main/java/com/google/gerrit/client/changes/ChangeConstants.properties
+++ b/gerrit-gwtui/src/main/java/com/google/gerrit/client/changes/ChangeConstants.properties
@@ -76,6 +76,7 @@ changePermalink = Permalink
 changeInfoBlockCanMerge = Can Merge
 changeInfoBlockCanMergeYes = Yes
 changeInfoBlockCanMergeNo = No
+changeInfoErrors = Errors
 
 includedInTableBranch = Branch Name
 includedInTableTag = Tag Name
diff --git a/gerrit-gwtui/src/main/java/com/google/gerrit/client/changes/ChangeDescriptionBlock.java b/gerrit-gwtui/src/main/java/com/google/gerrit/client/changes/ChangeDescriptionBlock.java
index c8b2a66..9748cd7 100644
--- a/gerrit-gwtui/src/main/java/com/google/gerrit/client/changes/ChangeDescriptionBlock.java
+++ b/gerrit-gwtui/src/main/java/com/google/gerrit/client/changes/ChangeDescriptionBlock.java
@@ -21,6 +21,8 @@ import com.google.gwt.user.client.ui.Composite;
 import com.google.gwt.user.client.ui.HorizontalPanel;
 import com.google.gwtexpui.globalkey.client.KeyCommandSet;
 
+import java.util.Set;
+
 public class ChangeDescriptionBlock extends Composite {
   private final ChangeInfoBlock infoBlock;
   private final CommitMessageBlock messageBlock;
@@ -36,8 +38,8 @@ public class ChangeDescriptionBlock extends Composite {
   }
 
   public void display(Change chg, Boolean starred, PatchSetInfo info,
-      final AccountInfoCache acc) {
-    infoBlock.display(chg, acc);
+      final AccountInfoCache acc, Set<Change.DependencyError> errors) {
+    infoBlock.display(chg, acc, errors);
     messageBlock.display(chg.getId(), starred, info.getMessage());
   }
 }
diff --git a/gerrit-gwtui/src/main/java/com/google/gerrit/client/changes/ChangeInfoBlock.java b/gerrit-gwtui/src/main/java/com/google/gerrit/client/changes/ChangeInfoBlock.java
index 3ffacc3..b390ec4 100644
--- a/gerrit-gwtui/src/main/java/com/google/gerrit/client/changes/ChangeInfoBlock.java
+++ b/gerrit-gwtui/src/main/java/com/google/gerrit/client/changes/ChangeInfoBlock.java
@@ -23,11 +23,19 @@ import com.google.gerrit.client.ui.ProjectLink;
 import com.google.gerrit.common.data.AccountInfoCache;
 import com.google.gerrit.reviewdb.client.Branch;
 import com.google.gerrit.reviewdb.client.Change;
+import com.google.gwt.dom.client.Document;
+import com.google.gwt.dom.client.Element;
+import com.google.gwt.dom.client.UListElement;
+import com.google.gwt.user.client.DOM;
 import com.google.gwt.user.client.ui.Composite;
 import com.google.gwt.user.client.ui.Grid;
+import com.google.gwt.user.client.ui.Widget;
 import com.google.gwt.user.client.ui.HTMLTable.CellFormatter;
+import com.google.gwt.user.client.ui.SimplePanel;
 import com.google.gwtexpui.clippy.client.CopyableLabel;
 
+import java.util.Set;
+
 public class ChangeInfoBlock extends Composite {
   private static final int R_CHANGE_ID = 0;
   private static final int R_OWNER = 1;
@@ -37,8 +45,9 @@ public class ChangeInfoBlock extends Composite {
   private static final int R_UPLOADED = 5;
   private static final int R_UPDATED = 6;
   private static final int R_STATUS = 7;
-  private static final int R_MERGE_TEST = 8;
-  private static final int R_CNT = 9;
+  private static final int R_ERRORS = 8;
+  private static final int R_MERGE_TEST = 9;
+  private static final int R_CNT = 10;
 
   private final Grid table;
 
@@ -59,6 +68,7 @@ public class ChangeInfoBlock extends Composite {
     initRow(R_UPLOADED, Util.C.changeInfoBlockUploaded());
     initRow(R_UPDATED, Util.C.changeInfoBlockUpdated());
     initRow(R_STATUS, Util.C.changeInfoBlockStatus());
+    initRow(R_ERRORS, Util.C.changeInfoErrors());
     if (Gerrit.getConfig().testChangeMerge()) {
       initRow(R_MERGE_TEST, Util.C.changeInfoBlockCanMerge());
     }
@@ -77,7 +87,8 @@ public class ChangeInfoBlock extends Composite {
     table.getCellFormatter().addStyleName(row, 0, Gerrit.RESOURCES.css().header());
   }
 
-  public void display(final Change chg, final AccountInfoCache acc) {
+  public void display(final Change chg, final AccountInfoCache acc,
+      Set<Change.DependencyError> errors) {
     final Branch.NameKey dst = chg.getDest();
 
     CopyableLabel changeIdLabel =
@@ -94,6 +105,23 @@ public class ChangeInfoBlock extends Composite {
     table.setText(R_UPLOADED, 1, mediumFormat(chg.getCreatedOn()));
     table.setText(R_UPDATED, 1, mediumFormat(chg.getLastUpdatedOn()));
     table.setText(R_STATUS, 1, Util.toLongString(chg.getStatus()));
+
+    if (errors != null && !errors.isEmpty()) {
+      table.setWidget(R_ERRORS, 1, new Widget() {
+        {
+          setElement(DOM.createElement("ul"));
+        }
+      });
+
+      for (Change.DependencyError error : errors) {
+        Element li = Document.get().createLIElement().cast();
+        li.setInnerText(error.toString().replace("_", " "));
+        table.getWidget(R_ERRORS, 1).getElement().appendChild(li);
+      }
+    } else {
+      table.getRowFormatter().setVisible(R_ERRORS, false);
+    }
+
     final Change.Status status = chg.getStatus();
     if (Gerrit.getConfig().testChangeMerge()) {
       if (status.equals(Change.Status.NEW) || status.equals(Change.Status.DRAFT)) {
diff --git a/gerrit-gwtui/src/main/java/com/google/gerrit/client/changes/ChangeScreen.java b/gerrit-gwtui/src/main/java/com/google/gerrit/client/changes/ChangeScreen.java
index 234f937..f6c1575 100644
--- a/gerrit-gwtui/src/main/java/com/google/gerrit/client/changes/ChangeScreen.java
+++ b/gerrit-gwtui/src/main/java/com/google/gerrit/client/changes/ChangeScreen.java
@@ -265,7 +265,8 @@ public class ChangeScreen extends Screen
     descriptionBlock.display(detail.getChange(),
         detail.isStarred(),
         detail.getCurrentPatchSetDetail().getInfo(),
-        detail.getAccounts());
+        detail.getAccounts(),
+        detail.getDependencyErrors());
     dependsOn.display(detail.getDependsOn());
     neededBy.display(detail.getNeededBy());
     approvals.display(detail);
diff --git a/gerrit-gwtui/src/main/java/com/google/gerrit/client/changes/PublishCommentScreen.java b/gerrit-gwtui/src/main/java/com/google/gerrit/client/changes/PublishCommentScreen.java
index 4705aad..ac3a4fc 100644
--- a/gerrit-gwtui/src/main/java/com/google/gerrit/client/changes/PublishCommentScreen.java
+++ b/gerrit-gwtui/src/main/java/com/google/gerrit/client/changes/PublishCommentScreen.java
@@ -274,7 +274,7 @@ public class PublishCommentScreen extends AccountScreen implements
   private void display(final PatchSetPublishDetail r) {
     setPageTitle(Util.M.publishComments(r.getChange().getKey().abbreviate(),
         patchSetId.get()));
-    descBlock.display(r.getChange(), null, r.getPatchSetInfo(), r.getAccounts());
+    descBlock.display(r.getChange(), null, r.getPatchSetInfo(), r.getAccounts(), null);
 
     if (r.getChange().getStatus().isOpen()) {
       initApprovals(r, approvalPanel);
diff --git a/gerrit-gwtui/src/main/java/com/google/gerrit/client/gerrit.css b/gerrit-gwtui/src/main/java/com/google/gerrit/client/gerrit.css
index 446b71d..33f76a6 100644
--- a/gerrit-gwtui/src/main/java/com/google/gerrit/client/gerrit.css
+++ b/gerrit-gwtui/src/main/java/com/google/gerrit/client/gerrit.css
@@ -929,6 +929,11 @@
   font-size: x-small;
 }
 
+.infoBlock td ul {
+  padding: 0 0 0 20px;
+  margin: 0;
+}
+
 
 .patchSetInfoBlock {
   margin-bottom: 10px;
-- 
1.7.8.3

